core-bundle/src/Util/SimpleTokenParser.php

```diff
@@ -0,0 +1,213 @@
+                array_pop($stack);
+                array_pop($ifStack);
+                $stack[] = !$currentIf && $stack[\count($stack) - 1] && $expression;
+                $ifStack[] = $currentIf || $expression;
+            } elseif (0 === strncmp($tag, '{else}', 6)) {
+                array_pop($stack);
+                array_pop($ifStack);
+                $stack[] = !$currentIf && $stack[\count($stack) - 1];
+                $ifStack[] = true;
+            } elseif (0 === strncmp($tag, '{endif}', 7)) {
+                array_pop($stack);
+                array_pop($ifStack);
+            } elseif ($current) {
+                $return .= $this->replaceTokens($tag, $data);
+            }
+        }
+
+        // Throw an exception if there is an error
+        if (1 !== \count($stack)) {
+            throw new \RuntimeException('Error parsing simple tokens');
+        }
+
+        return $return;
+    }
+
+    private function replaceTokens(string $subject, array $data): string
+    {
+        // Replace tokens
+        return preg_replace_callback(
+            '/##([^=!<>\s]+?)##/',
+            function (array $matches) use ($data) {
+                if (!\array_key_exists($matches[1], $data)) {
+                    if (null !== $this->logger) {
+                        $this->logger->log(LogLevel::INFO, sprintf('Tried to parse unknown simple token "%s".', $matches[1]));
+                    }
+
+                    return '##'.$matches[1].'##';
+                }
+
+                return $data[$matches[1]];
+            },
+            $subject
+        );
+    }
+
+    private function canUseExpressionLanguage(array $data): bool
+    {
+        foreach (array_keys($data) as $token) {
+            if (!preg_match('/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/', $token)) {
+                @trigger_error('Using tokens that are not valid PHP variables has been deprecated and will no longer work in Contao 5.0. Falling back to legacy token parsing.', E_USER_DEPRECATED);
+
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private function evaluateExpression(string $expression, array $data, bool $canUseExpressionLanguage): bool
+    {
+        if ($canUseExpressionLanguage) {
+            try {
+                return $this->expressionLanguage->evaluate($expression, $data);
+            } catch (\Exception $e) {
+                throw new \InvalidArgumentException($e->getMessage(), 0, $e);
+            }
+        }
+
+        // Legacy code
+        if (!preg_match('/^([^=!<>\s]+) *([=!<>]+)(.+)$/s', $expression, $matches)) {
+            return false;
+        }
+
+        [, $token, $operator, $value] = $matches;
+
+        if (!\array_key_exists($token, $data)) {
+            if (null !== $this->logger) {
+                $this->logger->log(LogLevel::INFO, sprintf('Tried to evaluate unknown simple token "%s".', $token));
```

Comment:
Wasn’t this logged to the Contao backend previously?

Reply:
It was taken from here:
https://github.com/contao/contao/blob/e7e0054dfc669f13e176217a1ed57282e2e7d987/core-bundle/src/Resources/contao/library/Contao/StringUtil.php#L572-L575